### ISSUE TITLE
x86_64-unknown-linux-musl build failed

### ISSUE THREAD

**Issue by zydxhs** [thecoshman/http#120]
src/util/os/non_windows_non_macos.rs:39:47

# mismatched type
```
error[E0308]: mismatched types
  --> src/util/os/non_windows_non_macos.rs:39:47
   |
39 |             let ok = unsafe { ioctl(dev_file, BLKGETSIZE, &mut block_count as *mut c_ulong) } == 0;
   |                                               ^^^^^^^^^^ expected `i32`, found `u64`
```
I change it to this:
```
let ok = unsafe { ioctl(dev_file, BLKGETSIZE as i32, &mut block_count as *mut c_ulong) } == 0;
```

# can't find ssl/crypto
```
= note: rust-lld: error: unable to find library -lssl
        rust-lld: error: unable to find library -lcrypto
```
OS: CentOS 7.8 x86_64
I have installed the openssl-devel.x86_64 and openssl-static.x86_64 library.
Compile Command:
```
CC=x86_64-linux-musl-gcc cargo build --release --target=x86_64-unknown-linux-musl
```

**Comment by nabijaczleweli** [comment]
Odd, `build.rs` should generate the right type for musl:
```rust
#[cfg(all(not(any(target_os = "windows", target_os = "macos")), not(target_env = "musl")))]
static IOCTL_REQUEST_TYPE: &str = "libc::c_ulong";

#[cfg(all(not(any(target_os = "windows", target_os = "macos")), target_env = "musl"))]
static IOCTL_REQUEST_TYPE: &str = "libc::c_int";
```
And, indeed, testing in an `x86_64-musl` Void chroot reveals that it does.

Can you build and run the following program with `rustc --target=x86_64-unknown-linux-musl` and post the output?
```rust
fn main() {
        #[cfg(target_env="")]     println!("empty");
        #[cfg(target_env="gnu")]  println!("gnu");
        #[cfg(target_env="msvc")] println!("msvc");
        #[cfg(target_env="musl")] println!("musl");
        #[cfg(target_env="sgx")]  println!("sgx");
}
```


As for the linker problems, this is outside the scope of this tracker, but either:
  * setting one of these variables that the `pkg-config` crate talks about: `OPENSSL_NO_PKG_CONFIG`, `PKG_CONFIG`, `OPENSSL_STATIC`, `OPENSSL_DYNAMIC`, `PKG_CONFIG_ALL_STATIC`, `PKG_CONFIG_ALL_DYNAMIC`, `PKG_CONFIG_PATH_x86_64-unknown-linux-musl`, `PKG_CONFIG_PATH_x86_64_unknown_linux_musl`, `HOST_PKG_CONFIG_PATH`, `PKG_CONFIG_PATH`, `PKG_CONFIG_LIBDIR_x86_64-unknown-linux-musl`, `PKG_CONFIG_LIBDIR_x86_64_unknown_linux_musl`, `HOST_PKG_CONFIG_LIBDIR`, `PKG_CONFIG_LIBDIR`, `PKG_CONFIG_SYSROOT_DIR_x86_64-unknown-linux-musl`, `PKG_CONFIG_SYSROOT_DIR_x86_64_unknown_linux_musl`, `HOST_PKG_CONFIG_SYSROOT_DIR`, `PKG_CONFIG_SYSROOT_DIR`, or
  * copying the requisite static `lib*.a` files from `/lib` into `target/release/deps/` (this is what I tend to use, even if a bit of a hack)

should work.

**Comment by nabijaczleweli** [comment]
Hm, weird, I can reproduce this, actually!

I've set up a cross-compilation config and building and running that program prints `musl` but doing the equivalent in the `build.rs` prints `gnu`; I'm assuming because build scripts are run on the host, but this is annoying nonetheless. I'll try to find a work-around for this.

**Comment by nabijaczleweli** [comment]
Okay, @zydxhs here's some actionable steps:

To fix the first thing, please pull `develop` (at least c2c6c633cc112ef6f4f9f352b1a7718cb831aadc).
To fix the second, can you try running `cp /usr/lib64/libcrypto.a /usr/lib64/libssl.a target/x86_64-unknown-linux-musl/release/deps/`?

**Comment by zydxhs** [comment]
I compiled successfully with the `messense/rust-musl-cross:x86_64-musl` Docker image.

**Comment by nabijaczleweli** [comment]
Top.

**Comment by zydxhs** [comment]
I used the compilation of `musl-cross-make` to install the cross-compilation tool, apt installed `libssl-dev`, and copied `libssl.a` and `libcrypto.a` to `target/x86_64-unknown-linux-musl/release/deps/`, Using the command ``CC= x86_64-Linux-musl-gcc cargo build -- release --target=x86_64-unknown-linux-musl" compile http, still failed.

What did I do wrong?

**Comment by nabijaczleweli** [comment]
With a clean `target/`, the following works on my Buster host (cc is Clang, but your toolchain is probably also fine):
```
nabijaczleweli@tarta:~/code/http$ rustup target add x86_64-unknown-linux-musl
nabijaczleweli@tarta:~/code/http$ CC=cc PKG_CONFIG=pkg-config cargo build --release --target=x86_64-unknown-linux-musl
nabijaczleweli@tarta:~/code/http$ ldd target/x86_64-unknown-linux-musl/release/http
        linux-vdso.so.1 (0x00007ffea1916000)
        libssl.so.1.1 => /lib/x86_64-linux-gnu/libssl.so.1.1 (0x00007f5298d4c000)
        libcrypto.so.1.1 => /lib/x86_64-linux-gnu/libcrypto.so.1.1 (0x00007f5298a63000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f5298a5e000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f5298a3d000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f529887c000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f5299314000)
```

If you post your error I might be able to diagnose further.

**Comment by zydxhs** [comment]
this is error info.
```
$ CC=x86_64-linux-musl-gcc PKG_CONFIG=pkg-config cargo build --release --target=x86_64-unknown-linux-musl
warning: file found to be present in multiple build targets: /home/zhuyadong/http/src/main.rs
   Compiling libc v0.2.81
   Compiling cc v1.0.66
   Compiling version_check v0.1.5
   Compiling autocfg v1.0.1
   Compiling rand_core v0.4.2
   Compiling autocfg v0.1.7
   Compiling cfg-if v0.1.10
   Compiling pkg-config v0.3.19
   Compiling log v0.4.11
   Compiling bitflags v1.2.1
   Compiling siphasher v0.2.3
   Compiling tinyvec_macros v0.1.0
   Compiling byteorder v1.3.4
   Compiling matches v0.1.8
   Compiling traitobject v0.1.0
   Compiling cfg-if v1.0.0
   Compiling lazy_static v1.4.0
   Compiling httparse v1.3.4
   Compiling openssl v0.10.32
   Compiling getrandom v0.1.15
   Compiling foreign-types-shared v0.1.1
   Compiling percent-encoding v1.0.1
   Compiling unicode-xid v0.0.4
   Compiling safemem v0.3.3
   Compiling memchr v2.3.4
   Compiling native-tls v0.2.6
   Compiling typeable v0.1.2
   Compiling language-tags v0.2.2
   Compiling quote v0.3.15
   Compiling gcc v0.3.55
   Compiling crc32fast v1.2.1
   Compiling unicode-width v0.1.8
   Compiling version_check v0.9.2
   Compiling nix v0.18.0
   Compiling openssl-probe v0.1.2
   Compiling antidote v1.0.0
   Compiling embed-resource v1.4.1
   Compiling ppv-lite86 v0.2.10
   Compiling serde v0.9.15
   Compiling adler v0.2.3
   Compiling strsim v0.8.0
   Compiling either v1.6.1
   Compiling same-file v1.0.6
   Compiling regex-syntax v0.6.21
   Compiling bitstring v0.1.1
   Compiling itoa v0.3.4
   Compiling modifier v0.1.0
   Compiling vec_map v0.8.2
   Compiling dtoa v0.4.6
   Compiling ansi_term v0.11.0
   Compiling percent-encoding v2.1.0
   Compiling lazysort v0.2.1
   Compiling os-str-generic v0.2.0
   Compiling trivial_colours v0.3.0
   Compiling xml-rs v0.8.3
   Compiling unicase v1.4.2
   Compiling num-traits v0.2.14
   Compiling miniz_oxide v0.4.3
   Compiling rand_core v0.3.1
   Compiling rand_jitter v0.1.4
   Compiling rand_pcg v0.1.2
   Compiling rand_chacha v0.1.1
   Compiling rand v0.6.5
   Compiling brotli-sys v0.3.2
   Compiling tinyvec v1.1.0
   Compiling openssl-sys v0.9.60
   Compiling bzip2-sys v0.1.9+1.0.8
   Compiling unicode-bidi v0.3.4
   Compiling unsafe-any v0.4.2
   Compiling foreign-types v0.3.2
   Compiling thread_local v1.0.1
   Compiling synom v0.11.3
   Compiling textwrap v0.11.0
   Compiling tabwriter v1.2.1
   Compiling unicase v2.6.0
   Compiling md6 v2.0.2
   Compiling https v1.10.0 (/home/zhuyadong/http)
   Compiling itertools v0.8.2
   Compiling walkdir v2.3.1
   Compiling cidr v0.1.1
   Compiling rand_hc v0.1.0
   Compiling rand_xorshift v0.1.1
   Compiling rand_isaac v0.1.1
   Compiling unicode-normalization v0.1.16
The following warnings were emitted during compilation:

warning: build/expando.c:2:10: 致命错误： openssl/opensslconf.h：没有那个文件或目录
warning:     2 | #include <openssl/opensslconf.h>
warning:       |          ^~~~~~~~~~~~~~~~~~~~~~~
warning: 编译中断。

error: failed to run custom build command for `openssl-sys v0.9.60`

Caused by:
  process didn't exit successfully: `/home/zhuyadong/http/target/release/build/openssl-sys-44f2674c93bf7221/build-script-main` (exit code: 101)
  --- stdout
  cargo:rustc-cfg=const_fn
  cargo:rerun-if-env-changed=X86_64_UNKNOWN_LINUX_MUSL_OPENSSL_LIB_DIR
  X86_64_UNKNOWN_LINUX_MUSL_OPENSSL_LIB_DIR unset
  cargo:rerun-if-env-changed=OPENSSL_LIB_DIR
  OPENSSL_LIB_DIR unset
  cargo:rerun-if-env-changed=X86_64_UNKNOWN_LINUX_MUSL_OPENSSL_INCLUDE_DIR
  X86_64_UNKNOWN_LINUX_MUSL_OPENSSL_INCLUDE_DIR unset
  cargo:rerun-if-env-changed=OPENSSL_INCLUDE_DIR
  OPENSSL_INCLUDE_DIR unset
  cargo:rerun-if-env-changed=X86_64_UNKNOWN_LINUX_MUSL_OPENSSL_DIR
  X86_64_UNKNOWN_LINUX_MUSL_OPENSSL_DIR unset
  cargo:rerun-if-env-changed=OPENSSL_DIR
  OPENSSL_DIR unset
  cargo:rerun-if-env-changed=OPENSSL_NO_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG_ALLOW_CROSS_x86_64-unknown-linux-musl
  cargo:rerun-if-env-changed=PKG_CONFIG_ALLOW_CROSS_x86_64_unknown_linux_musl
  cargo:rerun-if-env-changed=TARGET_PKG_CONFIG_ALLOW_CROSS
  cargo:rerun-if-env-changed=PKG_CONFIG_ALLOW_CROSS
  cargo:rerun-if-env-changed=PKG_CONFIG_x86_64-unknown-linux-musl
  cargo:rerun-if-env-changed=PKG_CONFIG_x86_64_unknown_linux_musl
  cargo:rerun-if-env-changed=TARGET_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG
  cargo:rerun-if-env-changed=OPENSSL_STATIC
  cargo:rerun-if-env-changed=OPENSSL_DYNAMIC
  cargo:rerun-if-env-changed=PKG_CONFIG_ALL_STATIC
  cargo:rerun-if-env-changed=PKG_CONFIG_ALL_DYNAMIC
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64-unknown-linux-musl
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64_unknown_linux_musl
  cargo:rerun-if-env-changed=TARGET_PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64-unknown-linux-musl
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64_unknown_linux_musl
  cargo:rerun-if-env-changed=TARGET_PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64-unknown-linux-musl
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64_unknown_linux_musl
  cargo:rerun-if-env-changed=TARGET_PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=SYSROOT
  cargo:rerun-if-env-changed=OPENSSL_STATIC
  cargo:rerun-if-env-changed=OPENSSL_DYNAMIC
  cargo:rerun-if-env-changed=PKG_CONFIG_ALL_STATIC
  cargo:rerun-if-env-changed=PKG_CONFIG_ALL_DYNAMIC
  cargo:rustc-link-lib=ssl
  cargo:rustc-link-lib=crypto
  cargo:rerun-if-env-changed=PKG_CONFIG
  cargo:rerun-if-env-changed=OPENSSL_STATIC
  cargo:rerun-if-env-changed=OPENSSL_DYNAMIC
  cargo:rerun-if-env-changed=PKG_CONFIG_ALL_STATIC
  cargo:rerun-if-env-changed=PKG_CONFIG_ALL_DYNAMIC
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64-unknown-linux-musl
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64_unknown_linux_musl
  cargo:rerun-if-env-changed=TARGET_PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64-unknown-linux-musl
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64_unknown_linux_musl
  cargo:rerun-if-env-changed=TARGET_PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64-unknown-linux-musl
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64_unknown_linux_musl
  cargo:rerun-if-env-changed=TARGET_PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR
  OPT_LEVEL = Some("3")
  TARGET = Some("x86_64-unknown-linux-musl")
  HOST = Some("x86_64-unknown-linux-gnu")
  CC_x86_64-unknown-linux-musl = None
  CC_x86_64_unknown_linux_musl = None
  TARGET_CC = None
  CC = Some("x86_64-linux-musl-gcc")
  CFLAGS_x86_64-unknown-linux-musl = None
  CFLAGS_x86_64_unknown_linux_musl = None
  TARGET_CFLAGS = None
  CFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  DEBUG = Some("false")
  CARGO_CFG_TARGET_FEATURE = Some("fxsr,sse,sse2")
  running: "x86_64-linux-musl-gcc" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "-m64" "-I" "/usr/include" "-Wall" "-Wextra" "-E" "build/expando.c"
  cargo:warning=build/expando.c:2:10: 致命错误： openssl/opensslconf.h：没有那个文件或目录
  cargo:warning=    2 | #include <openssl/opensslconf.h>
  cargo:warning=      |          ^~~~~~~~~~~~~~~~~~~~~~~
  cargo:warning=编译中断。
  exit code: 1

  --- stderr
  thread 'main' panicked at '
  Header expansion error:
  Error { kind: ToolExecError, message: "Command \"x86_64-linux-musl-gcc\" \"-O3\" \"-ffunction-sections\" \"-fdata-sections\" \"-fPIC\" \"-m64\" \"-I\" \"/usr/include\" \"-Wall\" \"-Wextra\" \"-E\" \"build/expando.c\" with args \"x86_64-linux-musl-gcc\" did not execute successfully (status code exit code: 1)." }

  Failed to find OpenSSL development headers.

  You can try fixing this setting the `OPENSSL_DIR` environment variable
  pointing to your OpenSSL installation or installing OpenSSL headers package
  specific to your distribution:

      # On Ubuntu
      sudo apt-get install libssl-dev
      # On Arch Linux
      sudo pacman -S openssl
      # On Fedora
      sudo dnf install openssl-devel

  See rust-openssl README for more information:

      https://github.com/sfackler/rust-openssl#linux
  ', /home/zhuyadong/.cargo/registry/src/mirrors.tuna.tsinghua.edu.cn-df7c3c540f42cdbd/openssl-sys-0.9.60/build/main.rs:141:13
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
error: build failed
```

**Comment by nabijaczleweli** [comment]
What do you get when you set `CC=cc` (or whatever the equivalent is for your system?).

**Comment by zydxhs** [comment]
```
$ cc --version
cc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by nabijaczleweli** [comment]
okay, but what happens if you set `CC=cc` and run the same thing again

**Comment by zydxhs** [comment]
command:
```
$ RUST_BACKTRACE=1 CC=cc PKG_CONFIG=pkg-config cargo build --release --target=x86_64-unknown-linux-musl
```
and the error:
```
   Compiling iron v0.6.1
error[E0308]: mismatched types
  --> src/util/os/non_windows_non_macos.rs:39:47
   |
39 |             let ok = unsafe { ioctl(dev_file, BLKGETSIZE, &mut block_count as *mut c_ulong) } == 0;
   |                                               ^^^^^^^^^^ expected `i32`, found `u64`
   |
help: you can convert an `u64` to `i32` and panic if the converted value wouldn't fit
   |
39 |             let ok = unsafe { ioctl(dev_file, BLKGETSIZE.try_into().unwrap(), &mut block_count as *mut c_ulong) } == 0;
   |                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
  --> src/util/os/non_windows_non_macos.rs:39:47
   |
39 |             let ok = unsafe { ioctl(dev_file, BLKGETSIZE, &mut block_count as *mut c_ulong) } == 0;
   |                                               ^^^^^^^^^^ expected `i32`, found `u64`
   |
help: you can convert an `u64` to `i32` and panic if the converted value wouldn't fit
   |
39 |             let ok = unsafe { ioctl(dev_file, BLKGETSIZE.try_into().unwrap(), &mut block_count as *mut c_ulong) } == 0;
   |                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
error: could not compile `https`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
```

then, I change 
```
let ok = unsafe { ioctl(dev_file, BLKGETSIZE.try_into().unwrap(), &mut block_count as *mut c_ulong) } == 0;
```
to 
```
let ok = unsafe { ioctl(dev_file, BLKGETSIZE as i32.try_into().unwrap(), &mut block_count as *mut c_ulong) } == 0;
```

this time, the error is:
```
  = note: /usr/bin/ld: /home/zhuyadong/http/target/x86_64-unknown-linux-musl/release/deps/libbrotli_sys-84c4deb38dae6039.rlib(brotli_bit_stream.o): undefined reference to symbol '__memset_chk@@GLIBC_2.3.4'
          //lib/x86_64-linux-gnu/libc.so.6: error adding symbols: DSO missing from command line
          collect2: error: ld returned 1 exit status
          

error: aborting due to previous error

error: build failed
```

The `ldd --version` output is:
```
ldd (Ubuntu GLIBC 2.27-3ubuntu1.2) 2.27
```

**Comment by nabijaczleweli** [comment]
Install Clang, set `CC=clang`, try again maybe?

**Comment by nabijaczleweli** [comment]
Released in [v1.10.1](https://github.com/thecoshman/http/releases/tag/v1.10.1).

**Comment by zydxhs** [comment]
Yeah, it's ok.
Thanks nabijaczleweli !
